common/mam/v1/mam.c

```diff
@@ -0,0 +1,188 @@
+  hamming(&curl, 0, HASH_LENGTH / 3, security);
+  mask(payload + offset, curl.state, HASH_LENGTH / 3, enc_curl);
+  offset += HASH_LENGTH / 3;
+
+  // encrypt signature to payload
+  curl_reset(&curl);
+  iss_curl_subseed(seed, payload + offset, start + index, &curl);
+  iss_curl_key(payload + offset, payload + offset, signature_length, &curl);
+  iss_curl_signature(payload + offset, enc_curl->state, 0, payload + offset,
+                     signature_length, &curl);
+  offset += signature_length;
+
+  // encrypt siblings number to payload
+  encode_long(siblings_number, payload + offset, enc_siblings_number_length);
+  offset += enc_siblings_number_length;
+
+  // encrypt siblings to payload
+  merkle_branch(payload + offset, merkle_tree, merkle_tree_length,
+                siblings_number + 1, index, leaf_count);
+  offset += siblings_number * HASH_LENGTH;
+
+  size_t to_mask = signature_length + enc_siblings_number_length +
+                   siblings_number * HASH_LENGTH;
+  mask(payload + offset - to_mask, payload + offset - to_mask, to_mask,
+       enc_curl);
+
+  curl_reset(enc_curl);
+
+  return payload_min_length;
+}
+
+int mam_parse(trit_t *payload, size_t payload_length, trit_t *message,
+              size_t *message_length, trit_t *side_key, size_t side_key_length,
+              trit_t *root, size_t *index, trit_t *next_root, size_t *security,
+              Curl *enc_curl) {
+  size_t offset = 0;
+
+  mam_init_encryption(side_key, side_key_length, root, enc_curl);
+
+  // decode index from payload
+  if (offset >= payload_length) return -1;
+  size_t enc_index_length;
+  *index =
+      decode_long(payload + offset, payload_length - offset, &enc_index_length);
+  offset += enc_index_length;
+
+  // decode message length from payload
+  if (offset >= payload_length) return -1;
+  size_t enc_message_length_length;
+  *message_length = decode_long(payload + offset, payload_length - offset,
+                                &enc_message_length_length);
+  offset += enc_message_length_length;
+
+  if (offset >= payload_length) return -1;
+  curl_absorb(enc_curl, payload, offset);
+
+  // decrypt next root from payload
+  unmask(next_root, payload + offset, HASH_LENGTH, enc_curl);
+  offset += HASH_LENGTH;
+
+  // decrypt message from payload
+  if (offset >= payload_length) return -1;
+  unmask(message, payload + offset, *message_length, enc_curl);
+  offset += *message_length;
+
+  // decrypt nonce from payload
+  if (offset >= payload_length) return -1;
+  unmask(payload + offset, payload + offset, HASH_LENGTH / 3, enc_curl);
+  offset += HASH_LENGTH / 3;
+
+  // get security back from state
+  if (offset >= payload_length) return -1;
+  trit_t hash[HASH_LENGTH];
+  memcpy(hash, enc_curl->state, HASH_LENGTH * sizeof(trit_t));
+  *security = signed_window(hash);
+  unmask(payload + offset, payload + offset, payload_length - offset, enc_curl);
+
+  if (*security == 0)
+    return -1;
+  else {
```

Comment:
Here `else` statement is not necessary since Line #155 already returns if `*security` equals to zero.

Reply:
Indeed, thanks